### PR TITLE
Add PyTorch implementation of BunDLeNet

### PR DESCRIPTION
**Pull request by vittoboa**
This PR adds the PyTorch implementation of BunDLe-Net:
- Conversion of existing TensorFlow code and models to PyTorch
- Addition of new tests
- Addition of GPU support (with _device_ parameter in _train_model_)
- Conversion of discrete model demo
- Addition of continuous PyTorch model demo (only the part related to BunDLe-Net, not the comparison with other embedding algorithms as that part wasn't converted yet to use the most recent version of the code)
- Conversion of _main.py_ and _continuous_main.py_

#### TensorFlow, what's left
- ``contrastive_loss`` in ``losses.py`` uses tensorflow, but it's not yet fully implemented or used so I didn't modify it and it's still the same.
- in ``bundlenet.py`` I left the ``get_config`` and ``from_config`` functions because it wasn't clear to me if they were directly related to tensorflow or not, because they weren't there initially. If they are related to tensorflow then they can be removed.

#### GPU
I have only been able to test the GPU in Google Colab, where it seems to work. When it's set on GPU, it trains more quickly and printing the model device ``next(model.parameters()).device`` or ``tensor.device`` correctly shows ``cuda:0``

#### Demos
I tried to keep the demos as similar as possible to the original ones, but I don't know if it's the way I am running them or what, but to make them work I have to add
```
import os
os.chdir('..')
```
at the beginning. And then change
``
data = np.load(f'../datasets/raw/rat_hippocampus/{rat_name}.npz')
``
to
``
data = np.load(f'datasets/raw/rat_hippocampus/{rat_name}.npz')
``
for the continuous. And 
``
data_path = 'datasets/raw/c_elegans/NoStim_Data.mat'
``
to
``
data_path = '../datasets/raw/c_elegans/NoStim_Data.mat'
``
in the discrete.
However, these changes weren't commited because I don't know if this problem is specific to the way I am running the demos. So these parts were just left the way they were.

### CODE REVIEW

File: ncmcm/visualisers/latent_space.py [hunk 1]
```diff
@@ -123,11 +123,9 @@ def _plot_ps(self, fig, ax, colors=None, **kwargs):
 
         for i in range(len(self.y) - 1):
             d = (self.y[i + 1] - self.y[i])
-            kwargs.setdefault('arrow_length_ratio', 0.1 / np.linalg.norm(d))
-            kwargs.setdefault('linewidths', 1)
             ax.quiver(self.y[i, 0], self.y[i, 1], self.y[i, 2],
                       d[0], d[1], d[2],
-                      color=color_dict[self.b[i]], **kwargs)
+                      color=color_dict[self.b[i]], arrow_length_ratio=0.1/np.linalg.norm(d), linewidths=1, **kwargs)
```
Comment (vittoboa) [inline ncmcm/visualisers/latent_space.py]: Note that this part reverts a change made in PR #20 because, at least on my side, that version made the plot_phase_space confusing/messy (both when using PyTorch and TensorFlow) because of the arrows.

Example:
phase_space

File: .github/workflows/test_main.yml [hunk 2]
```diff
@@ -38,6 +38,7 @@ jobs:
           python -m pip install --upgrade pip
           pip install pytest pytest-cov
           pip install -e .[full]
+          pip install -r verbose_requirements.txt
```
Comment (akshey-kumar) [inline .github/workflows/test_main.yml]: We are currently testing only with `pip install -e .` because we eventually want to release this as a pip package (PyPI). Hence, all dependencies should ideally be installed through `setup.py` . I created verbose_requirements.txt using a `pip freeze` just to have all the dependency versions that worked for us now, in case we have dependency issues at a later date.  In general, it is preferable if the package works with the latest stable versions of the dependencies, and we will ensure this as long as it is maintained. Hence `pip install -e .`  is sufficient.

File: demos/bundlenet_on_continuous_behaviour_rat_hippocampus_torch.ipynb
Comment (akshey-kumar) [inline demos/bundlenet_on_continuous_behaviour_rat_hippocampus_torch.ipynb:1]: I am quite glad to see the output of this. This convinces me that it was a very good decision to switch to PyTorch.

A bit of background: over the past weeks I wanted to update all dependencies to the latest versions, and was having quite a lot of trouble with tensorflow. Long story short, the results of BunDLe-Net in the continuous setting varied considerably between tf v2.15 and v2.16, the codebase and everything else remaining the same. I tried for quite some time to find what changed between the versions, but their release notes were barely helpful and there were several other bugs and inconvenient changes in API. Hence, I am quite glad to see that the manifold with your PyTorch qualitatively resembles the older runs.

File: ncmcm/bundlenet/bundlenet.py [hunk 3]
```diff
@@ -34,29 +47,38 @@ class BunDLeNet(Model):
...
+            nn.ReLU(),
+            nn.Linear(50, 30),
+            nn.ReLU(),
+            nn.Linear(30, 25),
+            nn.ReLU(),
+            nn.Linear(25, 10),
+            nn.ReLU(),
+            nn.Linear(10, latent_dim),
+            GaussianNoise(0.05),
+        )
```
Comment (akshey-kumar) [inline ncmcm/bundlenet/bundlenet.py:58]: Good job finding out that the normalization layer did not do anything! I verified that as well. Could you think of any replacement that results in the latent space being normalised? Basically, we don't want the method to 'cheat' by learning a manifold that is so small that it leads to a low dynamical loss even though it does not contain predictive information.

File: ncmcm/bundlenet/bundlenet.py [hunk 4]
```diff
@@ -34,29 +47,38 @@ class BunDLeNet(Model):
             discrete behavioural states
             For continuous-valued behaviours, this stipulates the number of
             behaviour variables
+
+        input_shape (tuple):
```
Comment (akshey-kumar) [inline ncmcm/bundlenet/bundlenet.py:35]: could you also add the other arguments as well to the docstring?

File: ncmcm/bundlenet/tests/test_bundlenet_continuous.py [hunk 5]
```diff
@@ -1,16 +1,17 @@
+import torch
 import numpy as np
 from ncmcm.bundlenet.utils import prep_data
 from ncmcm.bundlenet.bundlenet import BunDLeNet, train_model
 
 
 def test_bundlenet_training():
     X = np.random.rand(50, 10)
-    B = np.random.random(size=(50,2))
+    B = np.random.random(size=(50,2)).astype(np.float32)
```
Comment (akshey-kumar) [inline ncmcm/bundlenet/tests/test_bundlenet_continuous.py]: Why is this necessary?

File: ncmcm/visualisers/latent_space.py (hunk 1 above)
Comment (akshey-kumar) [inline ncmcm/visualisers/latent_space.py]: Doesn't this effectively lead to the same output? Instead of setting `arrow_length_ratio` in             `ax.quiver`, it is set in `kwargs.setdefault('arrow_length_ratio', 0.1 / np.linalg.norm(d))`. The latter version is more flexible since it allows the user to change the arrow size according the what suits the plot best.

File: ncmcm/bundlenet/bundlenet.py [hunk 6]
```diff
@@ -168,37 +192,44 @@ def train_model(x_train, b_train_1, model, b_type, gamma, learning_rate, n_epoch
...
         initialisation (str): 'pca_init' or 'best_of_5_init'
         validation_data: (x_test, b_test_1)
+        device (torch.device): Device where the model should be trained.
     Returns:
         numpy.ndarray: Array of loss values during training.
     """
-    train_dataset = tf_batch_prep(x_train, b_train_1)
-    optimizer = tf.keras.optimizers.Adam(learning_rate=learning_rate)
+    if device is None:
+        device = torch.device('cpu')
```
Comment (akshey-kumar) [inline ncmcm/bundlenet/bundlenet.py]: Maybe it's best to check if the machine has a gpu and use it as the default, if not, use the cpu? something along the lines of `device = torch.device("cuda" if torch.cuda.is_available() else "cpu")`

**Review by akshey-kumar — DISMISSED** [review]
You have done a good job overall! I am glad to see that you could send your a pull request along with your submission.

I have added a few minor comments that need to be addressed. In general, it would be great if you could add more documentation to your work wherever it is missing, or outdated. I have seen that you have mostly done this, but just got through it again to see if you have missed anything.

 If possible, please test if it works on a machine with a gpu. 
 
 Finally, feel free to add your name to the list of authors if you like:
 ```
"""
@authors:
Akshey Kumar
<your name goes here>
"""
```

File: verbose_requirements.txt [hunk 7]
```diff
@@ -67,6 +67,7 @@ statsmodels==0.14.2
 tensorboard==2.16.2
 tensorboard-data-server==0.7.2
 tensorflow==2.16.2
+torch==2.3.1
```
Comment (akshey-kumar) [inline verbose_requirements.txt]: I suggest the following:

add the pytorch to the setup.py file.
check if everything works fine ( with an `pip install .` on an empty virtual environment and without `pip install -r verbose_requirements`) 
if it does, do a pip freeze, so we have a record of what versions worked fine, should we have a problem at a later date.

File: ncmcm/bundlenet/bundlenet.py (hunk 4 above)
Comment (vittoboa) [reply ncmcm/bundlenet/bundlenet.py:35]: _latent_dim_ and _num_behaviour_ are actually already there, they're just not shown in the PR diff

File: ncmcm/bundlenet/bundlenet.py (hunk 6 above)
Comment (vittoboa) [reply ncmcm/bundlenet/bundlenet.py]: Yeah for sure, I updated it.
Then in this case, I also had to add ``device = next(model.parameters()).device`` to the demos to get the device where the model was created, so that even  for the evaluation the tensor can be created wherever the model is

File: .github/workflows/test_main.yml (hunk 2 above)
Comment (vittoboa) [reply .github/workflows/test_main.yml]: Removed :+1:

File: verbose_requirements.txt (hunk 7 above)
Comment (vittoboa) [reply verbose_requirements.txt]: Great, it does work! I saved the ``pip freeze`` in ``verbose_requirements_torch.txt``

File: ncmcm/bundlenet/tests/test_bundlenet_continuous.py (hunk 5 above)
Comment (vittoboa) [reply ncmcm/bundlenet/tests/test_bundlenet_continuous.py]: I made a [small change](https://github.com/NC-MCM/NC-MCM/pull/21/commits/4754ecf1112ec6ae9634948545226707acc41804) so that that’s not necessary anymore

File: ncmcm/bundlenet/bundlenet.py (hunk 3 above)
Comment (vittoboa) [reply ncmcm/bundlenet/bundlenet.py:58]: I have tried both using layer normalization _LayerNorm_ as well as batch normalization _BatchNorm1d_
For each, I have tried adding them:
• after every linear layer in tau
• after every linear layer in tau and after the linear layer of T_Y
• only after the last linear layer in tau
• only after the last linear layer in tau and after the linear layer of T_Y

The layer normalization didn’t work for me in all cases, it would just create a line for the continuous model/demo and a flat circle for the discrete one.
Batch normalization generally only worked similarly to the current version when placed only after the last linear layer of tau, but in general it looks slightly worse than without normalization. 

I updated the code and demos with _BatchNorm1d_ after the last linear layer of tau.

File: ncmcm/visualisers/latent_space.py (hunk 1 above)
Comment (vittoboa) [reply ncmcm/visualisers/latent_space.py]: On my end it didn't lead to the same output, when ``arrow_length_ratio`` is set directly in ``ax.quiver`` it's recalculated at each iteration. In the for loop, I got all the values of ``0.1/np.linalg.norm(d)`` and they ranged from 0.2 to 42

While if it is set in ``kwargs.setdefault`` it just has one value for all iterations. And at least on my end, the results seem very messy when using just one value.

For example:

_set directly in ax.quiver_
plot_phase_space_quiver


_kwargs.setdefault('arrow_length_ratio', 0.1 / np.linalg.norm(d))_
phase_space_kwargs

Comment (vittoboa) [reply ncmcm/visualisers/latent_space.py:129]: On my end it didn't lead to the same output, when ``arrow_length_ratio`` is set directly in ``ax.quiver`` it's recalculated at each iteration. In the for loop, I got all the values of ``0.1/np.linalg.norm(d)`` and they ranged from 0.2 to 42

While if it is set in ``kwargs.setdefault`` it just has one value for all iterations. And at least on my end, the results seem very messy when using just one value.

Some examples:
kwargs.setdefault('arrow_length_ratio', 0.1 / np.linalg.norm(d))
phase_space_setdefault

arrow_length_ratio = 5
phase_space_5


arrow_length_ratio = 10
phase_space_10

**Comment by vittoboa** [comment]
I have also updated the PR description with some additional notes

File: continuous_main.py [hunk 8]
```diff
@@ -36,7 +37,11 @@
 plt.legend()
 plt.show()
 
-y0_ = model.tau(x_[:,0]).numpy()
+device = next(model.parameters()).device
+
+model.eval()
+with torch.no_grad():
+    y0_ = model.tau(torch.tensor(x_[:, 0], dtype=torch.float, device=device)).cpu().numpy()
```
Comment (akshey-kumar) [inline continuous_main.py]: This added section of code that effectively serves to project the data seems a bit wordy to me. Could you simplify it, or wrap it up in some function so that the API remains simple to use? We want the API to be as high-level as possible (at the level a neuroscientist would like to use), without going into the deep learning implementation details. This would make it less confusing to the new user.

Try to mirror the older API (using TensorFlow) as closely as possible, or wrap up the other lines within already existing or, if necessary, a new function.

File: ncmcm/visualisers/latent_space.py (hunk 1 above)
Comment (akshey-kumar) [reply ncmcm/visualisers/latent_space.py]: That makes sense, thank you.

I would like to have a way by which one can control the size of arrows, and ideally other parameters of the quiver plot, directly through the plot_phase_space method, without having to go back to the src to change code. For example, in some plots I would like to have very small arrowheads, and in others larger ones, and in yet other, none at all.

File: continuous_main.py (hunk 8 above)
Comment (vittoboa) [reply continuous_main.py]: Unfortunately it can’t be simplified because everything is necessary (except ``with torch.no_grad()``, which just makes the code slightly faster by not computing the gradients as they are not necessary for inference). There is also no existing similar function where I could wrap the code, so I created a new function and made some tests for it

File: ncmcm/visualisers/latent_space.py (hunk 1 above)
Comment (vittoboa) [reply ncmcm/visualisers/latent_space.py]: Ah okey, then yeah, to control the quiver parameters we can use kwargs, so I brought it back.

Unfortunately the only argument to control the size of the arrow heads is ``arrow_length_ratio``, and it can only set the size of the head based on the arrow length, and not a size that it’s the same for all arrows.

So, for it I have set a hardcoded default value, because ``0.1/np.linalg.norm(d)`` frequently has values greater than 1, so the arrow head could end up being larger than the arrows.
